Blog Post: Fetch API for building servers (#790)
the-guild-org/the-guild-website: pages/blog/fetch-for-servers.mdx

  tags: ['fetch', 'whatwg', 'node', 'graphql', 'graphql-yoga'],
  authors: ['arda'],
  date: '2022-07-21',
  description: 'The new way of building servers in JavaScript',
  image:
    'https://user-images.githubusercontent.com/727224/104328926-56b0a700-54ba-11eb-9ad3-f6d2b971d55d.jpg',
  thumbnail:
    'https://user-images.githubusercontent.com/727224/104328777-29fc8f80-54ba-11eb-87d0-9982af1d8281.jpg'
}

export default withArticle(meta)

At the beginning of the year, we’ve launched GraphQL Yoga 2.0 - a server framework for GraphQL APIs.

While planning version 2.0 of Yoga, we were thinking about all the things that changed in the ecosystem and what developers using JavaScript expect from their server frameworks now and in the future.

One of the most powerful trend in the JS ecosystem was the proliferation of new environments and platforms that can run JS(Lambdas, Cloudflare Workers, Deno, Bun etc.). So we set up to build a single GraphQL server framework that could run on any of these platforms.

What we’ve found in the process was fascinating and we believe would change how any JS HTTP frameworks are being built, without any relation to GraphQL

> While Node.js is the most popular environment, multiple platforms can run JavaScript with different APIs. Usually, they have their own way of creating servers, but we saw they share a web standard that most people usually use for communicating with the backends from the client side, not the other way around. That’s WHATWG Fetch API that uses `Request`, `Response`, `ReadableStream`, etc to manage this on the server side, too.
> 

Reviewer: I'm not sure what you meant here so I suggested some changes by having some assumptions:

```suggestion
While Node.js is the most popular environment, multiple platforms can run JavaScript, usually, having their own way of creating servers, and using different APIs.

On the other hand, the JavaScript client-side has mostly migrated over the years to a common set of standards (`fetch`), independently of the underlying platform.

This is where we realized that WHATWG Fetch API that uses `Request`, `Response`, and `ReadableStream` could also be used on the server side to provide a unified API to run JavaScript HTTP servers everywhere.

```

## Why Fetch API?

> When you send a request from the client, you basically use `fetch(...requestArgs)` that uses `Request` object under the hood that contains all the details(headers, method, etc) and the data stream you need for that communication, then you take `Response` object that contains the connection stream and it processes it as you want with `.json`, `.arrayBuffer`, `.formData` or you can just access the stream itself by `.body` as `ReadableStream`

Reviewer: ```suggestion
When you send a request from the client, `fetch(...requestArgs)`  uses `Request` object under the hood that contains all the details(headers, method, etc) and the data stream you need for that communication. Then you take the `Response` object that contains the connection stream, and process it as you want with `.json()`, `.arrayBuffer()`, `.formData()` or just access the stream itself by `.body()` as `ReadableStream`
```

> On the server side, you can take `Request` object as a server and then process it with the same methods without dealing with the internals of your platform.

Reviewer: ```suggestion
On the server side, you can take `Request` object and process it with the same methods without dealing with the internals of your platform.
```

> Streaming responses like SSE, we just use `ReadableStream` and for Multipart requests (e.g. file uploads), we just take `FormData` which is exactly the same passed from the browser or any other client uses Fetch API.

Reviewer: ```suggestion
Streaming responses like SSE uses `ReadableStream`, and for Multipart requests (e.g. file uploads)uses  `FormData` which is exactly the same forwarded from the browser or any other client using Fetch API.
```

File Uploads; [https://github.com/dotansimha/graphql-yoga/blob/master/packages/common/src/plugins/requestParser/POSTMultipart.ts#L18](https://github.com/dotansimha/graphql-yoga/blob/master/packages/common/src/plugins/requestParser/POSTMultipart.ts#L18)

SSE; [https://github.com/dotansimha/graphql-yoga/blob/master/packages/common/src/plugins/resultProcessor/push.ts#L42](https://github.com/dotansimha/graphql-yoga/blob/master/packages/common/src/plugins/resultProcessor/push.ts#L42)

## What about Node.js?

> Currently, in the older LTS versions of Node.js, we don’t have an implementation of the Fetch API builtin. Furthermore, Node.js doesn’t use Web standard streams and the Fetch API in its `http` and `https` modules. That’s why we created the `@whatwg-node/fetch` package (previously known as `cross-undici-fetch` ) that fills in the gaps of different `fetch` implementations in all the LTS Node.js versions. Under the hood, `@whatwg-node/fetch` utilizes `undici` if available or otherwise falls back to using `node-fetch`, which you are probably already familiar with. In case `@whatwg-node/fetch` is imported in an environment that already has Fetch API built in like Cloudflare Workers, no ponyfills are added to your built application bundle.

Reviewer: ```suggestion
Currently, in the older LTS versions of Node.js, we don’t have an implementation of the Fetch API built in. Furthermore, Node.js doesn’t use Web standard streams and the Fetch API in its `http` and `https` modules. That’s why we created the `@whatwg-node/fetch` package (previously known as `cross-undici-fetch` ) that fills in the gaps of different `fetch` implementations in all the LTS Node.js versions. Under the hood, `@whatwg-node/fetch` utilizes `undici` if available or otherwise falls back to using `node-fetch`, which you are probably already familiar with. In case `@whatwg-node/fetch` is imported in an environment that already has Fetch API built in like Cloudflare Workers, no ponyfills are added to your built application bundle.
```

> - Ponyfill vs Polyfill

Reviewer: ```suggestion
**Ponyfill vs Polyfill**
```

  Polyfill patches the native parts of an environment while ponyfill just exports the “patched” stuff without touching the environment’s internals. We prefer pony filling because it prevents us from breaking other libraries and environmental functionalities.

## Is it possible to have a library that creates a cross-platform server?

> When rebuilding GraphQL Yoga from scratch, this was one of the most important features we wanted to implement. We wanted to create a GraphQL server library that can be integrated with different Node.js server frameworks, and other JS environments like CF Workers and Deno with a few additional lines of code.

Reviewer: ```suggestion
When rebuilding GraphQL Yoga from scratch, cross-platform support was one of the most important features we wanted to implement. We wanted to create a GraphQL server library that can be integrated with different Node.js server frameworks and other JS environments like CF Workers and Deno with a few additional lines of code.
```

// CF Workers
self.addEventListener('fetch', myServer)

// Next.js
export default myServer

// Deno
serve(myServer, {
  addr: ':4000'
})
```

### How does it look in real-life usage today?

You can check the GraphQL Yoga repository to see how we use this library;

[https://github.com/dotansimha/graphql-yoga/blob/no-more-node/packages/graphql-yoga/src/server.ts#L628](https://github.com/dotansimha/graphql-yoga/blob/no-more-node/packages/graphql-yoga/src/server.ts#L628)

And the simplicity of the integrations in our examples;

[https://github.com/dotansimha/graphql-yoga/tree/no-more-node/examples](https://github.com/dotansimha/graphql-yoga/tree/no-more-node/examples)

Finally how small the code is when we want to process the request and the response objects;

[https://github.com/dotansimha/graphql-yoga/tree/no-more-node/packages/graphql-yoga/src/plugins](https://github.com/dotansimha/graphql-yoga/tree/no-more-node/packages/graphql-yoga/src/plugins)

> There is literally nothing platform-specific in GraphQL Yoga, and this allows us to focus on creating a good GraphQL Server implementation for the entire GraphQL JS ecosystem.

## Node.js Server Frameworks, Routers & Middlewares

> Node.js was there in the JS community for years before the web standards were not as complete as today. We think Node.js inspired a lot of ideas on how server-side stuff can be easily managed with JavaScript by implementing all the Stream and HTTP APIs. But now we have Fetch, Web Streams, and so on as a web standard in the JavaScript ecosystem.

Reviewer: ```suggestion
Node.js solved issues years before the web standards could. However, we think that many server-side ideas inspired by Node.js can now easily be managed with JavaScript with Fetch, Web Streams, and so on as a web standard in the JavaScript ecosystem.
```

import withArticle from '../../ui/blog/article'

export const meta = {
  title: 'JavaScript runs everywhere, so should your servers - here is how',
  tags: ['fetch', 'whatwg', 'node', 'graphql', 'graphql-yoga'],
  authors: ['arda'],
  date: '2022-07-21',
  description: 'The new way of building servers in JavaScript',
  image:
    'https://user-images.githubusercontent.com/727224/104328926-56b0a700-54ba-11eb-9ad3-f6d2b971d55d.jpg',
  thumbnail:
    'https://user-images.githubusercontent.com/727224/104328777-29fc8f80-54ba-11eb-87d0-9982af1d8281.jpg'
}

export default withArticle(meta)

> 

Reviewer 2: ```suggestion
## TLDR
* We made GraphQL Yoga 2.0 **platform agnostic**, so it is able to run anywhere JavaScript can run (Cloudflare Workers, sdfsdf......) thanks to the new **Fetch API standard**
* We've created <ins>Ponyfills</ins> so it will work the same on older versions of Node that don't support the standard
* We've made a **new library** out of it so that any framework and app could achieve the same
* Help other frameworks in the ecosystem to migrate to this new library and standard
```

> At the beginning of the year, we’ve launched GraphQL Yoga 2.0 - a server framework for GraphQL APIs.

Reviewer 2: link to the blog post about graphql yoga

Reviewer: ```suggestion
At the beginning of the year, [we launched GraphQL Yoga 2.0](https://www.the-guild.dev/blog/announcing-graphql-yoga-v2) - a server framework for GraphQL APIs.
```

      })
    }
    return new Response('Nothing here!', {
      status: 404
    })
  }
})
```

There is another library called `itty-router` which can be used for Routing with Fetch API. You can see how simple it is to achieve “Routing” in a platform-independent way.

```tsx
import { Router } from 'itty-router'
import { createServerAdapter } from '@whatwg-node/server'

// now let's create a router (note the lack of "new")
const router = Router()

// GET collection index
router.get('/todos', () => new Response('Todos Index!'))

// GET item
router.get('/todos/:id', ({ params }) => new Response(`Todo #${params.id}`))

// POST to the collection (we'll use async here)
router.post('/todos', async (request) => {
  const content = await request.json()

  return new Response('Creating Todo: ' + JSON.stringify(content))
})

// 404 for everything else
router.all('*', () => new Response('Not Found.', { status: 404 }))

// attach the router "handle" to our server adapter
const myServer = createServerAdapter({
  handleRequest: router.handle
})

// Then use it in any environment
import { createServer } from 'http'

const httpServer = createServer(myServer)
httpServer.listen(4000)

> ```

Reviewer 2: ```suggestion
What NodeJS frameworks are you using today?
Maybe it is worth while to open a new issue on their repo and see if we could all help them to become platform agnostic, while still supporting older versions of Node, thanks to this new library!
```

import withArticle from '../../ui/blog/article'

export const meta = {
  title: 'JavaScript runs everywhere, so should your servers - here is how',
  tags: ['fetch', 'whatwg', 'node', 'graphql', 'graphql-yoga'],
  authors: ['arda'],
  date: '2022-07-21',
  description: 'The new way of building servers in JavaScript',
  image:
    'https://user-images.githubusercontent.com/727224/104328926-56b0a700-54ba-11eb-9ad3-f6d2b971d55d.jpg',
  thumbnail:

>     'https://user-images.githubusercontent.com/727224/104328777-29fc8f80-54ba-11eb-87d0-9982af1d8281.jpg'

Reviewer 3: We still need proper social images here.

import withArticle from '../../ui/blog/article'

export const meta = {

>   title: 'JavaScript runs everywhere, so should your servers - here is how',

Reviewer 3: ```suggestion
  title: 'JavaScript runs everywhere, so should your HTTP server - here is how',
```

I think this is more specific

Author: Hmm we need a new image then @Urigo @charlypoly What do you think?

Reviewer: I guess we can, let me know if you need a new set of images

Reviewer 2: I prefer the existing title.....

import withArticle from '../../ui/blog/article'

export const meta = {
  title: 'JavaScript runs everywhere, so should your servers - here is how',
  tags: ['fetch', 'whatwg', 'node', 'graphql', 'graphql-yoga'],
  authors: ['arda'],
  date: '2022-07-21',
  description: 'The new way of building servers in JavaScript',
  image: '/blog-assets/fetch-for-servers/cover.png',
  thumbnail: '/blog-assets/fetch-for-servers/thumbnail.png'
}

export default withArticle(meta)

## TLDR

> - We made GraphQL Yoga 2.0 **platform agnostic**, so it is able to run anywhere JavaScript can run (Cloudflare Workers, sdfsdf......) thanks to the new **Fetch API standard**

Author: Oh my bad. I forgot it while writing the draft 🤦

- We've created <ins>Ponyfills</ins> so it will work the same on older versions of Node that don't support the standard
- We've made a **new library** out of it so that any framework and app could achieve the same
- Help other frameworks in the ecosystem to migrate to this new library and standard

>   At the beginning of the year, [we launched GraphQL Yoga 2.0](https://www.the-guild.dev/blog/announcing-graphql-yoga-v2) - a server framework for GraphQL APIs.

Author: I actually do some introduction below but not sure what do you mean by a header actually?

This is where we realized that WHATWG Fetch API that uses [`Request`](https://developer.mozilla.org/en-US/docs/Web/API/Request), [`Response`](https://developer.mozilla.org/en-US/docs/Web/API/Response), and `ReadableStream` could also be used on the server side to provide a unified API to run JavaScript HTTP servers everywhere.

## Why Fetch API?

When you send a request from the client, `fetch(...requestArgs)` uses [`Request`](https://developer.mozilla.org/en-US/docs/Web/API/Request) object under the hood that contains all the details(headers, method, etc) and the data stream you need for that communication. Then you take the [`Response`](https://developer.mozilla.org/en-US/docs/Web/API/Response) object that contains the connection stream, and process it as you want with `.json()`, `.arrayBuffer()`, `.formData()` or just access the stream itself by `.body()` as `ReadableStream`

On the server side, you can take [`Request`](https://developer.mozilla.org/en-US/docs/Web/API/Request) object and process it with the same methods without dealing with the internals of your platform.

> Streaming responses like SSE uses `ReadableStream`, and for Multipart requests (e.g. file uploads)uses `FormData` which is exactly the same forwarded from the browser or any other client using Fetch API.

Author: I mean you can get the exactly same `FormData` on the server side with `request.formData`.

After a few iterations, it was clear to us that this is certainly possible and finally shipped this as part of GraphQL Yoga v2.

GraphQL Yoga instance itself can be used directly as a request listener that you pass to Express’s `app.use`, Node’s native `http.createServer`, Next.js functions. and same for other non-Node.js environments; we just pass GraphQL Yoga as an event listener for CF Workers’ `self.addEventListener('fetch', yoga)`.

As we already mentioned in the “Why Fetch API?” part, the server library itself doesn’t need to care about the platform’s connection-specific details like Node’s `IncomingMessage` and `ServerResponse` or Next.js’s `NextApi.Request` objects. You are now able to focus on your server implementation details by consuming a “universal standard” [`Request`](https://developer.mozilla.org/en-US/docs/Web/API/Request) and returning “another standard” [`Response`](https://developer.mozilla.org/en-US/docs/Web/API/Response)

> As we realized, how well this works out for our users, we decided that we need to bring this to the next level and extract that logic into a standalone library called `@whatwg-node/server` . You simply provide your request handler that has a single [`Request`](https://developer.mozilla.org/en-US/docs/Web/API/Request) parameter and expects you to return a [`Response`](https://developer.mozilla.org/en-US/docs/Web/API/Response) instance. The generated request handler instance can be integrated with regular Node HTTP servers, Fastify, Koa, Deno, CF Workers, Next.js, etc with a few lines of code.

Reviewer 2: ```suggestion
As we realized how well this works out for our users, we decided that we need to bring this to the next level and extract that logic into a standalone library called [`@whatwg-node/server` ](https://github.com/ardatan/whatwg-node). 

You simply provide your request handler that has a single [`Request`](https://developer.mozilla.org/en-US/docs/Web/API/Request) parameter and expects you to return a [`Response`](https://developer.mozilla.org/en-US/docs/Web/API/Response) instance. The generated request handler instance can be integrated with regular Node HTTP servers, Fastify, Koa, Deno, CF Workers, Next.js, etc with a few lines of code.
```

// CF Workers
self.addEventListener('fetch', myServer)

// Next.js
export default myServer

// Deno
serve(myServer, {
  addr: ':4000'
})
```

### How does it look in real-life usage today?

You can check the GraphQL Yoga repository to see how we use this library;

[https://github.com/dotansimha/graphql-yoga/blob/no-more-node/packages/graphql-yoga/src/server.ts#L628](https://github.com/dotansimha/graphql-yoga/blob/no-more-node/packages/graphql-yoga/src/server.ts#L628)

And the simplicity of the integrations in our examples;

[https://github.com/dotansimha/graphql-yoga/tree/no-more-node/examples](https://github.com/dotansimha/graphql-yoga/tree/no-more-node/examples)

Finally how small the code is when we want to process the request and the response objects;

[https://github.com/dotansimha/graphql-yoga/tree/no-more-node/packages/graphql-yoga/src/plugins](https://github.com/dotansimha/graphql-yoga/tree/no-more-node/packages/graphql-yoga/src/plugins)

> There is literally nothing platform-specific in GraphQL Yoga, and this allows us to focus on creating a good GraphQL Server implementation for the entire GraphQL JS ecosystem.

## Node.js Server Frameworks, Routers & Middlewares

> Node.js solved issues years before the web standards could. However, we think that many server-side ideas inspired by Node.js can now easily be managed with JavaScript with Fetch, Web Streams, and so on as a web standard in the JavaScript ecosystem.

Reviewer 2: ```suggestion
Node.js solved issues years before the web standards could. However, we think that many server-side ideas inspired by Node.js can now easily be managed with JavaScript with Fetch, Web Streams and other web standards in the JavaScript ecosystem.
```

> There are many mature libraries such as Fastify, Koa, Express, and Hapi that are implemented only for Node.js which Fetch API doesn’t have in its ecosystem for now. The experience with those libraries in the current era of Node.js taught us a lot about how the server can be designed but maybe it is time to reduce the environment-specific APIs in the JS ecosystem.

Reviewer 2: ```suggestion
There are many mature libraries such as Fastify, Koa, Express, and Hapi that are implemented only for Node.js without using the Fetch API Standard. The experience with those libraries in the current era of Node.js taught us a lot about how the server can be designed but maybe it is time to reduce the environment-specific APIs in the JS ecosystem.
```

The major reason for using a server framework is usually “Routing” then “Middlewares” so the question is “Why cannot we just have that with Fetch API?”

> You can basically achieve routing like below however it looks a bit unsafe.

Reviewer 2: ```suggestion
You can basically achieve routing like below, however it looks a bit unsafe.
```

      status: 404
    })
  }
})
```

There is another library called `itty-router` which can be used for Routing with Fetch API. You can see how simple it is to achieve “Routing” in a platform-independent way.

```tsx
import { Router } from 'itty-router'
import { createServerAdapter } from '@whatwg-node/server'

// now let's create a router (note the lack of "new")
const router = Router()

// GET collection index
router.get('/todos', () => new Response('Todos Index!'))

// GET item
router.get('/todos/:id', ({ params }) => new Response(`Todo #${params.id}`))

// POST to the collection (we'll use async here)
router.post('/todos', async (request) => {
  const content = await request.json()

  return new Response('Creating Todo: ' + JSON.stringify(content))
})

// 404 for everything else
router.all('*', () => new Response('Not Found.', { status: 404 }))

// attach the router "handle" to our server adapter
const myServer = createServerAdapter({
  handleRequest: router.handle
})

// Then use it in any environment
import { createServer } from 'http'

const httpServer = createServer(myServer)
httpServer.listen(4000)
```

## What NodeJS frameworks are you using today?

> Maybe it is worth while to open a new issue on their repo and see if we could all help them to become platform agnostic, while still supporting older versions of Node, thanks to this new library!

Reviewer 2: ```suggestion
Maybe it is worth while to open a new issue on their repo and see if we could all help them to become platform agnostic, while still supporting older versions of Node, thanks to this new library.

Please try it out and give us feedback [on the repo](https://github.com/ardatan/whatwg-node)!
```
